### ISSUE TITLE
Make Deployment Reconciler leader-aware

### ISSUE THREAD

**Issue by nachocano** [google/knative-gcp#1466]
**Problem**
While updating dependencies, it seems that eventing is enabling leader-election for the control-plane. 
Currently, all our reconcilers are leader-aware, but the Deployment one. 
We should make it leader-aware so that we can support HA. It is currently disabled in our controller.

From the logs: 
"2020/07/15 20:57:51 *deployment.Reconciler is not leader-aware, all reconcilers must be leader-aware to enable fine-grained leader election."

Tried implementing it using a new Deployment genreconciler, but that doesn't work, as genreconciler is going to resync everything in a bucket when promote is called. 
We need to implement the LeaderAwareFuncs ourselves as in: 
https://github.com/knative/pkg/blob/25f2aa6443803868e6562dca5e57626f2309302d/webhook/configmaps/controller.go#L53-L59

**[Persona:](https://github.com/knative/eventing/blob/master/docs/personas.md)** 
Developer

**Exit Criteria**
Leader-aware deployment reconciler to enable HA

**Comment by nachocano** [comment]
The reconciler keeps on getting called, with any deployment in the cluster, and things keep on being restarted.

`{"level":"info","ts":"2020-07-16T16:47:45.298Z","logger":"controller.github.com-google-knative-gcp-pkg-reconciler-deployment.Reconciler","caller":"deployment/deployment.go:58","msg":"Annotating the controller","commit":"c3225c0","knative.dev/traceid":"d32a2aed-fa0b-446e-81a6-69abb2f5d74b","knative.dev/key":"cloud-run-events/webhook","name":"webhook","namespace":"cloud-run-events"}
{"level":"info","ts":"2020-07-16T16:47:45.334Z","logger":"controller.github.com-google-knative-gcp-pkg-reconciler-deployment.Reconciler","caller":"controller/controller.go:474","msg":"Reconcile succeeded. Time taken: 35.625265ms","commit":"c3225c0","knative.dev/traceid":"d32a2aed-fa0b-446e-81a6-69abb2f5d74b","knative.dev/key":"cloud-run-events/webhook"}
{"level":"info","ts":"2020-07-16T16:47:45.334Z","logger":"controller.github.com-google-knative-gcp-pkg-reconciler-deployment.Reconciler","caller":"deployment/deployment.go:58","msg":"Annotating the controller","commit":"c3225c0","knative.dev/traceid":"2a5b681b-e663-4cdc-b1e5-384fd6ebf389","knative.dev/key":"gke-system/istio-ingress","name":"istio-ingress","namespace":"gke-system"}
{"level":"info","ts":"2020-07-16T16:47:45.343Z","logger":"controller.github.com-google-knative-gcp-pkg-reconciler-deployment.Reconciler","caller":"controller/controller.go:474","msg":"Reconcile succeeded. Time taken: 47.591784ms","commit":"c3225c0","knative.dev/traceid":"43bd94f1-4da9-4ced-a9ac-bfacdc1bf3ad","knative.dev/key":"cloud-run-events/controller"}
{"level":"info","ts":"2020-07-16T16:47:45.343Z","logger":"controller.github.com-google-knative-gcp-pkg-reconciler-deployment.Reconciler","caller":"deployment/deployment.go:58","msg":"Annotating the controller","commit":"c3225c0","knative.dev/traceid":"af04aa88-7f7f-4f48-acf0-3931d15dedfd","knative.dev/key":"kube-system/kube-dns-autoscaler","name":"kube-dns-autoscaler","namespace":"kube-system"}
{"level":"info","ts":"2020-07-16T16:47:45.401Z","logger":"controller.github.com-google-knative-gcp-pkg-reconciler-deployment.Reconciler","caller":"controller/controller.go:474","msg":"Reconcile succeeded. Time taken: 66.997043ms","commit":"c3225c0","knative.dev/traceid":"2a5b681b-e663-4cdc-b1e5-384fd6ebf389","knative.dev/key":"gke-system/istio-ingress"}
{"level":"info","ts":"2020-07-16T16:47:45.401Z","logger":"controller.github.com-google-knative-gcp-pkg-reconciler-deployment.Reconciler","caller":"deployment/deployment.go:58","msg":"Annotating the controller","commit":"c3225c0","knative.dev/traceid":"9a8d9cc5-e9f0-4fb5-bcea-40adab84e4ce","knative.dev/key":"kube-system/l7-default-backend","name":"l7-default-backend","namespace":"kube-system"}
{"level":"info","ts":"2020-07-16T16:47:45.491Z","logger":"controller.github.com-google-knative-gcp-pkg-reconciler-deployment.Reconciler","caller":"controller/controller.go:474","msg":"Reconcile succeeded. Time taken: 148.182138ms","commit":"c3225c0","knative.dev/traceid":"af04aa88-7f7f-4f48-acf0-3931d15dedfd","knative.dev/key":"kube-system/kube-dns-autoscaler"}
{"level":"info","ts":"2020-07-16T16:47:45.491Z","logger":"controller.github.com-google-knative-gcp-pkg-reconciler-deployment.Reconciler","caller":"deployment/deployment.go:58","msg":"Annotating the controller","commit":"c3225c0","knative.dev/traceid":"1f764a89-95a9-4bc4-8fcd-4ad18802b35a","knative.dev/key":"kube-system/fluentd-gke-scaler","name":"fluentd-gke-scaler","namespace":"kube-system"}
{"level":"info","ts":"2020-07-16T16:47:45.527Z","logger":"controller.github.com-google-knative-gcp-pkg-reconciler-deployment.Reconciler","caller":"controller/controller.go:474","msg":"Reconcile succeeded. Time taken: 126.329886ms","commit":"c3225c0","knative.dev/traceid":"9a8d9cc5-e9f0-4fb5-bcea-40adab84e4ce","knative.dev/key":"kube-system/l7-default-backend"}`


I cannot find a way that it is only called with our own deployment. 
I think the solution is to actually have a secret reconciler, which will detect a change in the particular secret. And it should get our deployment and patch it if necessary.

**Comment by nachocano** [comment]
> I cannot find a way that it is only called with our own deployment.
> I think the solution is to actually have a secret reconciler, which will detect a change in the particular secret. And it should get our deployment and patch it if necessary.

Tried this but it has the same problem. 
Synced with Matt, and just updated the issue description. We should not use genreconciler, but instead implement LeaderAwareFuncs ourselves.

**Comment by nachocano** [comment]
Removing the milestone as disabling HA unblocks our update dependencies. 
We need to re-think the patching of the secret. Problem mentioned here: https://github.com/google/knative-gcp/pull/1465#issuecomment-659904919
Some initial code in: https://github.com/nachocano/knative-gcp/tree/leader-aware/pkg/reconciler/deployment

**Comment by nachocano** [comment]
The Deployment reconciler is some hack we needed to so that as soon as we patched the control-plane secret, we restarted the pod. Otherwise it could take a while. See https://kubernetes.io/docs/concepts/configuration/secret/#mounted-secrets-are-updated-automatically, and in practice we saw it took ~1 min.

Re-assigning this to Security as it's more related to that area than just Sources, and folks might be able to rethink about this secret hack.
